Expand CSS style guide to cover more conventions and recommendations (#40228)
mdn/content: files/en-us/mdn/writing_guidelines/code_style_guide/css/index.md

Before diving in and writing huge chunks of CSS, plan your styles carefully. What general styles are going to be needed, what different layouts do you need to create, what specific overrides need to be created, and are they reusable? Above all, you need to try to **avoid too much overriding**. If you keep finding yourself writing styles and then cancelling them again a few rules down, you probably need to rethink your strategy.

### Using modern CSS features

> You can use new features once every major browser — Chrome, Edge, Firefox, and Safari — supports them.

Author: We are not talking about documentation criteria. We are talking about using CSS features as if they can be used in production---for example the CSS of a Web API example application.

> You can use new features once every major browser — Chrome, Edge, Firefox, and Safari — supports them.

Reviewer: The text needs to say that it is allowed to use the new features for demonstrating them but not to be used as regular features in unrelated contexts.

### Don't use resets

For maximum control over CSS across platforms, many people used to use CSS resets to remove all styles before then building things back up themselves. This certainly has its merits, but especially in the modern world, CSS resets can be overkill, resulting in a lot of extra time spent reimplementing things that weren't completely broken in the first place, such as default margins and list styles.

### Formal syntax and pseudocode

> Formal syntax is more front and central for CSS docs than other docs, and readers tend to understand them to some degree, so they are not forbidden even outside of the "Formal syntax" section. However, if your code is not syntactically well-formed CSS, do not mark it as CSS. Either mark your code block as `plain`, or use the `CSSSyntaxRaw` macro to render complete formal syntax. Syntax errors in `css` code blocks results in the code being unparsable by static checkers, confuses readers who expect to see valid CSS code, and may even result in nonsensical syntax highlighting.

Author: ```suggestion
Formal syntax is an integral part of MDN's CSS documentation (as an example, see the [Formal syntax](/en-US/docs/Web/CSS/background-image#formal_syntax) section on the `background-image` property page). Because a lot of developers are familiar with syntax in this format, it's acceptable to write pseudocode in a formal-syntax-like fashion in descriptions and examples. However, formal syntax is not syntactically well-formed CSS, and should, therefore, not be marked as such. Either mark your code block as `plain`, or use the `CSSSyntaxRaw` macro to render the complete formal syntax.
```

For maximum control over CSS across platforms, a lot of people used to use CSS resets to remove every style, before then building things back up themselves. This certainly has its merits, but especially in the modern world, CSS resets can be an overkill, resulting in a lot of extra time spent reimplementing things that weren't completely broken in the first place, like default margins, list styles, etc.

### Pseudocode and formal syntax

> Formal syntax is more front and central for CSS docs than other docs, and readers tend to understand them to some degree, so they are not forbidden. However, if your code is not syntactically well-formed CSS, do not mark it as CSS. Either mark your code block as `plain`, or use the `CSSSyntaxRaw` macro to render complete formal syntax.

Author: "Syntax" section should show plausible CSS code (not necessarily "valid" due to the CSS functions exception) and not CSS formal syntax. "Formal syntax" section should use `CSSSyntaxRaw` but that's yet another topic.

We do have some cases where formal syntax is used outside of the "Formal syntax" section, which is fine for me.

> Formal syntax is more front and central for CSS docs than other docs, and readers tend to understand them to some degree, so they are not forbidden. However, if your code is not syntactically well-formed CSS, do not mark it as CSS. Either mark your code block as `plain`, or use the `CSSSyntaxRaw` macro to render complete formal syntax.

Author: I'm not just talking about the "Formal syntax" section though. Often we just randomly throw out formal syntax in description or examples.

> Furthermore, a single value is not syntactically well-formed CSS. CSS code at least requires a property and its value. If you are documenting the `rgb()` function, write this:

Author: Syntax code is not a code example, and this guideline doesn't apply in the first place. "Syntax sections" and "Code examples" are two parallel parts according to https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures

> Furthermore, a single value is not syntactically well-formed CSS. CSS code at least requires a property and its value. If you are documenting the `rgb()` function, write this:

Author: > I guess what we want to say here is that outside of "Syntax", if your pseudocode is not valid CSS, mark it as `plain`?

I've clarified that this section does not apply to the syntax code block.

```css example-good
color: rgb(31 41 59);
color: rgb(31 41 59 / 26%);
```

Don't use this style:

```css example-bad
rgb(31 41 59);
rgb(31 41 59 / 26%);
```

## Animations

> ### Keyframe selectors

Author: H2 headings are broad CSS feature categorizations; H3 headings are specific features or suggestions

> ### Keyframe selectors

Author: I understand about this general editorial rule, but IMO occasionally we need to lend to structural consistencies elsewhere. For example, the "Examples" section already breaks this rule, because we have the conventions of (1) an H2 called "Examples" (2) each example having its H3 with a semantic description of what it's doing.

When specifying keyframes, the `0%` and `100%` selectors can also be written as `from` and `to`. If a `@keyframes` rule _only contains_ these two selectors, use `from` and `to` instead of `0%` and `100%`. This makes your code more semantic.

```css example-bad
@keyframes example {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
```

```css example-good
@keyframes example {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
```

On the other hand, if your `@keyframes` rule contains more than just the start and end frames, use the `0%` and `100%` selectors for uniformity.

```css example-good
@keyframes example {
  0% {
    opacity: 0;
  }
  50% {
    opacity: 0.8;
  }
  100% {
    opacity: 1;
  }
}
```

## Cascade, properties, and selectors

> ### Controlling specificity

Author: I did not intend to rename this section; I tried out your suggestion but decided I didn't like it, and I forgot to revert this part.

If possible, avoid surprises for controlling specificity, such as using `:where()` or duplicating selectors. Consider:

- Changing the order of declarations

> - Rearranging properties such that they don't override each other

Author: It's not necessarily "unintentionally", for example you may have a default-state and a special-state.

> - Using ID selectors

Author: (Same comment below) I intend for this to only apply within limited scenarios, namely the one I gave for when ID selectors themselves are acceptable:

> ID selectors are acceptable if unique handles are required elsewhere; for example, form controls.

For example, if you have `<button class="foo" id="bar" />` anyway because you need a handle to `bar` in JavaScript, then it's not a big deal if you have styles for `.foo` and styles for `#bar` and take advantage of specificities.

> - Using ID selectors

Author: My personal philosophy is that unique classes are bad practices. Classes should be inherently specify behavior and not particular elements. If a class by definition can only apply to a single element on the page then it should be an ID.

> - Using ID selectors

Reviewer: For small codes in live samples, it doesn't matter. However, on a large scale, you always want to reuse an element or component. If an ID were used in style sheets, you would also have to update the CSS and HTML. Most of the sites are not handcrafted. We need to promote good practices for broader perspectives.

As IDs are for identification, they are most suitable for identification in JavaScript and for anchoring purposes.

  ```css example-bad
  margin-top: 1em;
  margin-right: 2em;
  margin-bottom: 1em;
  margin-left: 2em;
  ```

  With this:

  ```css example-good
  margin: 1em 2em;
  ```

- If _some_ constituent properties of a shorthand property are specified to a non-default value, the use of the shorthand property is optional. Both of these are acceptable:

  ```css example-good
  margin-top: 1em;
  margin-bottom: 1em;
  ```

  ```css example-good
  margin: 1em 0;
  ```

- Use the shortest shorthand syntax available. Write this:

  ```css example-good
  margin: 1em;
  ```

  Avoid these:

  ```css example-bad
  margin: 1em 1em;
  margin: 1em 1em 1em 1em;
  ```

- Write shorthand properties in the canonical order. Write this:

  ```css example-good
  border: 1px solid red;
  ```

  Don't write this:

  ```css example-bad
  border: solid red 1px;
  ```

- If you use the shorthand property, make sure none of its constituent properties are specified at the same time, because the longhand is ignored. Avoid this:

  ```css example-bad
  margin-top: 1em;
  margin: 2em; /* Oops, margin-top is ignored */

>   ```

Author: Yeah but this is still very confusing, which is why we should either use the shorthand or combination of longhands but not a mix of both.

### Use class selectors

Generally, prefer class selectors (and use `class` instead of `id` in your HTML). They can be composed: multiple elements can use the same class, and the same class can be used for multiple elements.

```css example-good
.footnote {
  /* ... */
}
```

```css example-bad
#footnote {
  /* ... */
}
```

> ID selectors are acceptable if unique handles are required elsewhere; for example, form controls. They are also an acceptable device for controlling specificity.

Reviewer 2: ```suggestion
ID selectors are acceptable if unique handles are required elsewhere, for example, for form controls.

Use classes for styling, and reserve IDs for non-CSS purposes, such as for use in JavaScript or for linking to unique page anchors (`<a href="#section1">`).
```

The `:is()`, `:where()`, and `:not()` pseudo-classes accept complex selector lists. Use them to shorten your selector.

Write this:

```css example-good
input:not(:checked, :disabled) {
  /* ... */
}
```

Don't write this:

```css example-bad
input:not(:checked):not(:disabled) {
  /* ... */
}
```

## Colors

### Choosing a notation

Generally, if the specific color palette is not a concern, default to using common named colors. For example, use `black` instead of `rgb(0 0 0)` or `#000000`, and `green` instead of `chartreuse`.

If a specific color is needed, default to using the `rgb()` notation. `hsl()` and other functions should only be used where the particular representation has a meaning (for example, a color wheel or a gradient). Hexadecimal notation is terser but may be less readable; it is interchangeable with `rgb()` depending on which one is more convenient for you.

Whatever color function you use, always use the modern syntax: `rgb(31 41 59 / 26%)` and not the legacy comma-separated syntax. Always use the function without the `a` suffix (`rgb` instead of `rgba`), because it's shorter and doesn't require changing the name if you later decide to add an alpha channel.

> When using the hexadecimal notation, always use the six (or eight) digit version to avoid cognitive load: `#aabbcc` instead of `#abc`.

Author: I would like to leave this freedom for the writer, especially because certain color palettes may come in hex and the alpha is the hardest to convert to decimal

}
```

### Ordering

Generally, when two declarations target the same element(s), the one with higher specificity should come later in the stylesheet.

```css example-good
button {
  color: blue;
}

.my-form button {
  color: red;
}
```

Within a declaration, prefer to have related properties (such as sizing, positioning, color, etc.) located together. Custom properties should be declared at the top, which makes the code more readable if the custom property is to be used within the same block.

### Empty lines

Empty lines between declaration blocks are recommended. You can remove them if consecutive declarations are highly related, such as variations of the same utility class.

Empty lines between properties should be used sparingly. Only use them when each group of properties form a clear semantic block.

### Shorthand properties

> - If _every_ constituent property of a shorthand property is specified to a non-default value, use the shorthand property instead of the constituent properties. This makes your code shorter and easier to read. Replace this:
> 
>   ```css example-bad
>   margin-top: 1em;
>   margin-right: 2em;
>   margin-bottom: 1em;
>   margin-left: 2em;
>   ```

Reviewer 2: ```suggestion
nits
```suggestion
- If _every_ constituent property of a shorthand property is assigned a non-default value, use the shorthand property instead of the constituent longhand properties. This makes your code shorter and easier to read. 

Replace these longhand properties:

  ```css example-bad
  margin-top: 1em;
  margin-right: 2em;
  margin-bottom: 1em;
  margin-left: 2em;
  ```
```

>   With this:
> 
>   ```css example-good
>   margin: 1em 2em;
>   ```

Reviewer 2: ```suggestion
  with its corresponding shorthand:
  
  ```css example-good
  margin: 1em 2em;
  ```
```

> - If _some_ constituent properties of a shorthand property are specified to a non-default value, the use of the shorthand property is optional. Both of these are acceptable:

Reviewer 2: ```suggestion
- If only _some_ constituent properties of a shorthand property are assigned a non-default value, the use of the shorthand property is optional. Both of these are acceptable:
```

  ```css example-good
  margin-top: 1em;
  margin-bottom: 1em;
  ```

  ```css example-good
  margin: 1em 0;
  ```

- Use the shortest shorthand syntax available. Write this:

  ```css example-good
  margin: 1em;
  ```

  Avoid these:

  ```css example-bad
  margin: 1em 1em;
  margin: 1em 1em 1em 1em;
  ```

- Write shorthand properties in the canonical order. Write this:

  ```css example-good
  border: 1px solid red;
  ```

  Don't write this:

  ```css example-bad
  border: solid red 1px;
  ```

> - If you use the shorthand property, make sure none of its constituent properties are specified at the same time, because the longhand is ignored. Avoid this:

Reviewer 2: ```suggestion
- If you use a shorthand property, make sure none of its constituent longhand properties are specified at the same time because the longhand properties will be ignored. Avoid this:
```

  ```css example-bad
  margin-top: 1em;
  margin: 2em; /* Oops, margin-top is ignored */
  ```

### Use class selectors

> Generally, prefer class selectors (and use `class` instead of `id` in your HTML). They can be composed: multiple elements can use the same class, and the same class can be used for multiple elements.

Reviewer 2: ```suggestion
Generally, prefer [class selectors](/en-US/docs/Web/CSS/Class_selectors) (and use `class` instead of `id` in your HTML). They can be composed: multiple elements can use the same class, and the same class can be used for multiple elements.
```

```css example-good
.footnote {
  /* ... */
}
```

```css example-bad
#footnote {
  /* ... */
}
```

ID selectors are acceptable if unique handles are required elsewhere; for example, form controls. They are also an acceptable device for controlling specificity.

### Old pseudo-element selectors

The `::before`, `::after`, `::first-letter`, and `::first-line` pseudo-elements can also be written with single colons. Avoid the single-colon syntax because they are conceptually pseudo-elements and not pseudo-classes.

### Complex selector lists

> The `:is()`, `:where()`, and `:not()` pseudo-classes accept complex selector lists. Use them to shorten your selector.

Reviewer 2: individual pseudo-classes can be linked, adding link for complex selectors
```suggestion
The `:is()`, `:where()`, and `:not()` pseudo-classes accept [complex selector lists](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#complex_selector). Use them to shorten your selector.
```

Write this:

```css example-good
input:not(:checked, :disabled) {
  /* ... */
}
```

Don't write this:

```css example-bad
input:not(:checked):not(:disabled) {
  /* ... */
}
```

## Casing

> By default, all identifiers should be lowercase. This applies to selectors, functions, and keywords. Custom identifiers should use kebab-case, such as `--custom-property` or `my-animation`.

Author: Class names are not CSS. They are HTML that happen to be referenced in CSS. As such they are defined in the HTML style guide

> The exceptions are the keyword values defined in SVG, which for historical reasons are camelCase, and should be written as such to enhance readability. They include: `currentColor`, {{cssxref("text-rendering")}} values, {{cssxref("shape-rendering")}} values, {{cssxref("pointer-events")}} values, and {{cssxref("color-interpolation-filters")}} values.

Reviewer 2: ```suggestion
Exceptions include keyword values defined in SVG, which for historical reasons are [camelCase](/en-US/docs/Glossary/Camel_case), and should be written as such to enhance readability. These keywords include: `currentColor`, {{cssxref("text-rendering")}} values, {{cssxref("shape-rendering")}} values, {{cssxref("pointer-events")}} values, and {{cssxref("color-interpolation-filters")}} values.
```

## Colors

### Choosing a notation

> Generally, if the specific color palette is not a concern, default to using common named colors. For example, use `black` instead of `rgb(0 0 0)` or `#000000`, and `green` instead of `chartreuse`.

Reviewer: ```suggestion
Try to use [named colors](en-US/docs/Web/CSS/named-color#standard_colors). for example, use `black` instead of `rgb(0 0 0)` or `#000000`. If the specific color palette is not a concern, then use familiar, recognizable color names instead of their shades, for example, `green` instead of `chartreuse`.
```

> Generally, if the specific color palette is not a concern, default to using common named colors. For example, use `black` instead of `rgb(0 0 0)` or `#000000`, and `green` instead of `chartreuse`.

Author: The "If the specific color palette is not a concern" condition applies to both "use named colors" and "use familiar names". If you need to use a specific palette (for example, branded colors) then you can't use named colors anyway. In other words, `chartreuse` is equally as bad as `#000000` according to this guideline.

If a specific color is needed, default to using the `rgb()` notation. `hsl()` and other functions should only be used where the particular representation has a meaning (for example, a color wheel or a gradient). Hexadecimal notation is terser but may be less readable; it is interchangeable with `rgb()` depending on which one is more convenient for you.

> Whatever color function you use, always use the modern syntax: `rgb(31 41 59 / 26%)` and not the legacy comma-separated syntax. Always use the function without the `a` suffix (`rgb` instead of `rgba`), because it's shorter and doesn't require changing the name if you later decide to add an alpha channel.

Author: > Do we have an opinion about `0.26` vs `.26`?

The former is prevalent and I prefer it.

> Whatever color function you use, always use the modern syntax: `rgb(31 41 59 / 26%)` and not the legacy comma-separated syntax. Always use the function without the `a` suffix (`rgb` instead of `rgba`), because it's shorter and doesn't require changing the name if you later decide to add an alpha channel.

Reviewer 2: ```suggestion
Whatever color function you use, always use the modern syntax (`rgb(31 41 59 / 0.26)`), not the legacy comma-separated one. Always use the function without the `a` suffix (`rgb` instead of `rgba`), because it's shorter and doesn't require changing the name if you later decide to add an alpha channel.
```

When using the hexadecimal notation, always use the six (or eight) digit version to avoid cognitive load: `#aabbcc` instead of `#abc`.

### Color parameters

> For consistency, all parameters should use numbers by default, instead of percentages or degrees. This includes the alpha channel too. However, if the specific representation is meaningful (for example, animations, gradients, or calculations), then use the suitable type.

Reviewer 2: ```suggestion
For consistency, all parameters should use numbers by default instead of percentages or degrees. This also applies to the alpha channel. However, if a specific representation is meaningful (for example, in animations, gradients, or calculations), use the suitable type.
```

If the alpha channel is `1`, omit it. Write `rgb(31 41 59)` instead of `rgb(31 41 59 / 1)`.

### Choosing colors

> In addition to the recommendation of using common named colors, your color palette should meet our [accessibility guidelines](/en-US/docs/Web/Accessibility/Guides/Colors_and_Luminance). In particular, if the colors distinguish elements (such as a "red box" and a "blue box"), make sure that the colors are distinguishable by people with color vision deficiency.

Author: The contrast requirement is already covered by the accessibility guidelines. I'm just pointing out a specific case of using colors to designate meaning because it's not as common in real web applications.

> In addition to the recommendation of using common named colors, your color palette should meet our [accessibility guidelines](/en-US/docs/Web/Accessibility/Guides/Colors_and_Luminance). In particular, if the colors distinguish elements (such as a "red box" and a "blue box"), make sure that the colors are distinguishable by people with color vision deficiency.

Reviewer: Still, it would be beneficial to emphasize color contrast. Because people do raise issues just to report that the 4.5:1 contrast is not being followed in some example. Sometimes reviewers also insist on improving the contrast.

I've added https://developer.mozilla.org/en-US/docs/Web/Accessibility/Guides/Understanding_WCAG/Perceivable/Color_contrast link to my suggestin above.

  Replace these longhand properties:

  ```css example-bad
  margin-top: 1em;
  margin-right: 2em;
  margin-bottom: 1em;
  margin-left: 2em;
  ```

  with their corresponding shorthand:

  ```css example-good
  margin: 1em 2em;
  ```

- If only _some_ constituent properties of a shorthand property are assigned a non-default value, the use of the shorthand property is optional. Both of these are acceptable:

  ```css example-good
  margin-top: 1em;
  margin-bottom: 1em;
  ```

  ```css example-good
  margin: 1em 0;
  ```

- Use the shortest shorthand syntax available. Write this:

  ```css example-good
  margin: 1em;
  ```

  Avoid these:

  ```css example-bad
  margin: 1em 1em;
  margin: 1em 1em 1em 1em;
  ```

- Write shorthand properties in the [canonical order](/en-US/docs/Glossary/Canonical_order). Write this:

  ```css example-good
  /* width style color */
  border: 1px solid red;
  ```

  Don't write this:

  ```css example-bad
  border: solid red 1px;
  ```

- For every shorthand, either use it or its longhand constituents, and never a mix of both, because the overriding relationship is complex and bug-prone. Avoid these:

>   ```css example-bad

Reviewer: `border-bottom-width` is not part of the `border` shorthand [ref](https://developer.mozilla.org/en-US/docs/Web/CSS/border#formal_syntax). To avoid future disputes, need to use a different example...

  margin-top: 1em;
  margin: 2em; /* Oops, margin-top is ignored */

  border-width: 1px;
  border-bottom-width: 5px; /* Overrides one border's width *only* */
  ```

### Use class selectors

Generally, prefer [class selectors](/en-US/docs/Web/CSS/Class_selectors) (and use `class` instead of `id` in your HTML). They can be composed: multiple elements can use the same class, and the same class can be used for multiple elements.

```css example-good
.footnote {
  /* ... */
}
```

```css example-bad
#footnote {
  /* ... */
}
```

> ID selectors are acceptable if unique handles are required elsewhere, for example, for form controls.
> 
> Use classes for styling, and reserve IDs for non-CSS purposes, such as for use in JavaScript or for linking to unique page anchors (`<a href="#section1">`). In the case where the use of ID is justified, you can take advantage of it for [controlling specificity](#controlling_specificity).

Author: ```suggestion
Use classes for styling, and reserve IDs for non-CSS purposes, such as for use in JavaScript or for linking to unique page anchors (`<a href="#section1">`). In the case where the use of ID is justified, you can use it as a selector, potentially to [control specificity](#controlling_specificity).
```

The `::before`, `::after`, `::first-letter`, and `::first-line` [pseudo-elements](/en-US/docs/Web/CSS/Pseudo-elements) can also be written with single colons (like `:before`). Avoid the single-colon syntax because the syntax is discouraged and they could be misidentified as [pseudo-classes](/en-US/docs/Web/CSS/Pseudo-classes) (`:hover`) by readers.

### Complex selector lists

The `:is()`, `:where()`, and `:not()` pseudo-classes accept [complex selector lists](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#complex_selector). Use them to shorten your selector.

Write this:

```css example-good
input:not(:checked, :disabled) {
  /* ... */
}
```

Don't write this:

```css example-bad
input:not(:checked):not(:disabled) {
  /* ... */
}
```

## Casing

By default, all identifiers should be lowercase. This applies to selectors, functions, and keywords. Custom identifiers should use [kebab-case](/en-US/docs/Glossary/Kebab_case), such as `--custom-property` or `my-animation`. See the [HTML style guide](/en-US/docs/MDN/Writing_guidelines/Code_style_guide/HTML#casing_convention_on_mdn) for casing conventions of HTML IDs and classes which are referenced as CSS selectors.

> Exceptions include keyword values defined in SVG, which for historical reasons are [camelCase](/en-US/docs/Glossary/Camel_case), and should be written as such to enhance readability. These keywords include: [`currentColor`](/en-US/docs/Web/CSS/color_value#currentcolor_keyword), {{cssxref("text-rendering")}} values, {{cssxref("shape-rendering")}} values, {{cssxref("pointer-events")}} values, and {{cssxref("color-interpolation-filters")}} values.

Author: If you really want to add commas it would be "which, for historical reasons, are camelCase". I also don't think "be in camelCase" and "be camelCase" would be unfamiliar to anyone; "camelCase" can practically be an adjective just like "camelCase identifiers".

## Colors

### Choosing a notation

Generally, if the specific color palette is not a concern, default to using common named colors. For example, use `black` instead of `rgb(0 0 0)` or `#000000`, and `green` instead of `chartreuse`.

If a specific color is needed, default to using the `rgb()` notation. `hsl()` and other functions should only be used where the particular representation has a meaning (for example, a color wheel or a gradient). Hexadecimal notation is terser but may be less readable; it is interchangeable with `rgb()` depending on which one is more convenient for you.

> Whatever color function you use, always use the modern syntax (`rgb(31 41 59 / 0.26)`), not the legacy comma-separated one. Always use the function without the `a` suffix (`rgb` instead of `rgba`), because it's shorter and doesn't require changing the name if you later decide to add an alpha channel.

Author: ```suggestion
Whatever color function you use, always use the modern syntax (`rgb(31 41 59 / 0.26)`), not the legacy comma-separated one. Always use the function without the `a` suffix (`rgb` instead of `rgba`), because it's shorter and doesn't require changing the name if you later decide to add or remove the alpha channel.
```

When using the hexadecimal notation, always use the six (or eight) digit version to avoid cognitive load: `#aabbcc` instead of `#abc`.

### Color parameters

For consistency, all parameters should use numbers by default instead of percentages or degrees. This also applies to the alpha channel. However, if a specific representation is meaningful (for example, in animations, gradients, or calculations), use the suitable type in the context.

If the alpha channel is `1`, omit it. Write `rgb(31 41 59)` instead of `rgb(31 41 59 / 1)`.

### Choosing colors

> In addition to the recommendation of using common named colors, your color palette should meet our [accessibility guidelines](/en-US/docs/Web/Accessibility/Guides/Colors_and_Luminance). In particular, if the colors distinguish elements (such as a "red box" and a "blue box"), ensure that the colors are distinguishable to people with color vision deficiency. Aim for at least 4.5:1 [contrast ratio](/en-US/docs/Web/Accessibility/Guides/Understanding_WCAG/Perceivable/Color_contrast) (WCAG AA) between text and background.

Reviewer: My bad, in another tab, the older version of the doc was open.

}
```

## Fonts

### Specifying font families

> When specifying a font family, always add a generic family name as the last fallback. This ensures that if the specified font is not available, the browser displays a more suitable font than the default font. [Web-safe fonts](/en-US/docs/Learn_web_development/Core/Text_styling/Fundamentals#web_safe_fonts) are exempt from this rule.

Reviewer 2: ```suggestion
When specifying a font family, always add a [generic font family](/en-US/docs/Web/CSS/font-family#generic-name) name as the last fallback. This ensures that if the specified font is not available, the browser displays a more suitable fallback font. [Web-safe fonts](/en-US/docs/Learn_web_development/Core/Text_styling/Fundamentals#web_safe_fonts) are exempt from this rule.
```

body {
  font-family: "Helvetica";
}
```

```css example-good
body {
  /* The "sans-serif" family is not needed because Arial is a web-safe font */
  font-family: "Helvetica", "Arial";
}

math {
  font-family: "Latin Modern Math", "STIX Two Math", math;
}
```

### Specifying font weights

Prefer keyword values such as `normal` and `bold`, and relative weights such as `bolder` and `lighter`. Only use number values where the specific weight is desired. You should always replace `400` with `normal` and `700` with `bold`, unless for consistency with other similar declarations.

## Lengths

### Use flexible/relative units

For maximum flexibility over the widest possible range of devices, default to relative units like `em`, `rem`, percentages, and viewport units (if you want them to vary depending on viewport width) for all lengths. You can read some more about this in our [guide to CSS values and units](/en-US/docs/Learn_web_development/Core/Styling_basics/Values_and_units#relative_length_units).

Write this:

```css example-good
margin: 0.5em;
max-width: 50%;
```

Avoid this:

```css example-bad
margin: 20px;
max-width: 500px;
```

## Media queries

### Range syntax

> Use the modern range syntax instead of `min-` and `max-`. The former allows specifying exclusive ranges, allows simultaneously specifying upper and lower bounds, and is overall more concise and readable.

Reviewer: ```suggestion
Use the modern range syntax instead of `min-` and `max-`. The former allows specifying exclusive ranges, allows simultaneous specification of upper and lower bounds, and is overall more concise and readable.
```

Author: They sound equivalent to me.

Reviewer: Is there a way the reviewer can resolve the conversation?  Delete doesn't resolve/remove it...

```css example-good
@media (width >= 480px) {
  /* ... */
}
@media (600px < height < 900px) {
  /* ... */
}
```

```css example-bad
@media (min-width: 480px) {
  /* ... */
}
@media (min-height: 600px) and (max-height: 900px) {
  /* ... */
}
```

This principle extends to HTML attributes, such as the `media` attribute of `<link>` elements.

> If you have different alternative styles, be especially careful with your media queries. Remember that `width` and `height` can be fractional values; make sure that with every value, there is one and only one alternative style in effect.

Reviewer: ```suggestion
If you have different alternative styles, be especially careful with your media queries. Remember that `width` and `height` can be fractional values; ensure that with every value, there is only one alternative style in effect.
```

}
```

## Fonts

### Specifying font families

When specifying a font family, always add a [generic font family](/en-US/docs/Web/CSS/font-family#generic-name) name as the last fallback. This ensures that if the specified font is not available, the browser displays a more suitable fallback font. [Web-safe fonts](/en-US/docs/Learn_web_development/Core/Text_styling/Fundamentals#web_safe_fonts) are exempt from this rule.

```css example-bad
body {
  font-family: "Helvetica";
}
```

```css example-good
body {
  /* The "sans-serif" family is not needed because Arial is a web-safe font */
  font-family: "Helvetica", "Arial";
}

math {
  font-family: "Latin Modern Math", "STIX Two Math", math;
}
```

### Specifying font weights

> Prefer keyword values such as `normal` and `bold`, and relative weights such as `bolder` and `lighter`. Only use number values where the specific weight is desired. You should always replace `400` with `normal` and `700` with `bold`, except when declaring ranges with variable fonts, or for consistency with other similar declarations.

Reviewer: ```suggestion
Prefer keyword values such as `normal` and `bold`, and relative weights such as `bolder` and `lighter`. Use only number values when the specific weight is required. You should always replace `400` with `normal` and `700` with `bold`, except when declaring ranges with variable fonts, or for consistency with other similar declarations.
```

Author: "Only use" means "This is the only case where you can use". "Use only" means "Use nothing other than".

## Lengths

### Use flexible/relative units

For maximum flexibility over the widest possible range of devices, default to relative units like `em`, `rem`, percentages, and viewport units (if you want them to vary depending on viewport width) for all lengths. You can read some more about this in our [guide to CSS values and units](/en-US/docs/Learn_web_development/Core/Styling_basics/Values_and_units#relative_length_units).

Write this:

```css example-good
margin: 0.5em;
max-width: 50%;
```

Avoid this:

```css example-bad
margin: 20px;
max-width: 500px;
```

## Media queries

### Range syntax

Use the modern range syntax instead of `min-` and `max-`. The former allows specifying exclusive ranges, allows simultaneously specifying upper and lower bounds, and is overall more concise and readable.

```css example-good
@media (width >= 480px) {
  /* ... */
}

> @media (600px < height < 900px) {

Reviewer: ```suggestion

@media (600px < height < 900px) {
```

Author: I personally would prefer to squeeze as many lines as possible... And this is not real CSS code anyway, so I'm just stacking examples together.

  /* ... */
}
```

```css example-bad
@media (min-width: 480px) {
  /* ... */
}

> @media (min-height: 600px) and (max-height: 900px) {

Reviewer: ```suggestion

@media (min-height: 600px) and (max-height: 900px) {
```

  /* ... */
}
```

> This principle extends to HTML attributes, such as the `media` attribute of `<link>` elements.

Author: ```suggestion
This principle extends to non-CSS usage of media queries, such as the [`media`](/en-US/docs/Web/HTML/Reference/Elements/link#media) attribute of `<link>` elements or {{domxref("window.matchMedia()")}}.
```

> If you have different alternative styles, be especially careful with your media queries. Remember that `width` and `height` can be fractional values; make sure that with every value, there is one and only one alternative style in effect.

Author: ```suggestion
If you have different alternative styles selected by media thresholds, be especially careful with your media queries. Remember that `width` and `height` can be fractional values; make sure that with every value, there is one and only one alternative style in effect.
```

}
```

## Strings

> Wherever quotes are optional in the CSS syntax, use them, and use double quotes. Do this:

Author: I didn't write this; this was already there. I just made the implication about `font-family` explicit because before it didn't have an example of that.

The CSS spec says: "To avoid mistakes in escaping, it is recommended to quote font family names that contain white space, digits, or punctuation characters other than hyphens". So we either add quotes under these circumstances only, or we require quotes everywhere. The latter sounds significantly simpler as a rule, and in general I like rules that say "always do X" rather than "omit X if Y", because if later Y doesn't hold anymore (for example, you change "MonteCarlo" to "Monte Carlo"), then you suddenly need to add X back.

> Wherever quotes are optional in the CSS syntax, use them, and use double quotes. Do this:

Reviewer: Déjà vu moment for me. Weren't we discussing about distinguishing between string and literal values using q uotes? I think it was about API or CSS def lists in value sections.

> Wherever quotes are optional in the CSS syntax, use them, and use double quotes. Do this:

Author: Yes, it was about whether to use ``- `"constant"` `` or ``- `constant` `` in value DLs for API and HTTP, but that's not very related.

}
```

## Strings

Wherever quotes are optional in the CSS syntax, use them, and use double quotes. Do this:

```css example-good
[data-vegetable="liquid"] {
  background-image: url("../../media/examples/lizard.png");
  font-family: "Helvetica", "Arial";
}
```

> Don't do this, because the types of characters allowed are more limited and sometimes lead to subtle syntax errors:

Reviewer: ```suggestion
Don't do the following, because the types of characters allowed are more limited and sometimes lead to subtle syntax errors:
```
